### PR TITLE
[ML][Inference] Adding classification_weights to ensemble models

### PR DESCRIPTION
**Pull request by benwtrent**
classification_weights are a way to allow models to
prefer specific classification results over others
this might be advantageous if classification value
probabilities are a known quantity and can improve
model error rates.

### CODE REVIEW

**Review by tveasey — APPROVED** [review]
LGTM

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/langident/LangIdentNeuralNetwork.java [hunk 1]
```diff
@@ -156,11 +146,6 @@ public TargetType targetType() {
         return TargetType.CLASSIFICATION;
     }
 
-    @Override
-    public List<String> classificationLabels() {
-        return LANGUAGE_NAMES;
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/langident/LangIdentNeuralNetwork.java]: Consider adding `LANGUAGE_NAMES` to `toXContent` and the parser. `Tree` includes the labels in its xcontent representation

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceHelpers.java [hunk 2]
```diff
@@ -20,27 +21,40 @@
...
-        }
-        int[] sortedIndices = IntStream.range(0, probabilities.size())
-            .boxed()
-            .sorted(Comparator.comparing(probabilities::get).reversed())
-            .mapToInt(i -> i)
-            .toArray();
+    /**
+     * @return Tuple of the highest scored index and the top classes
+     */
+    public static Tuple<Integer, List<ClassificationInferenceResults.TopClassEntry>> topClasses(List<Double> probabilities,
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceHelpers.java]: Is there no test for this method?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/TrainedModel.java [hunk 3]
```diff
@@ -6,21 +6,14 @@
...
 
-import java.util.List;
 import java.util.Map;
 
 public interface TrainedModel extends NamedXContentObject, NamedWriteable, Accountable {
 
-    /**
-     * @return List of featureNames expected by the model. In the order that they are expected
-     */
-    List<String> getFeatureNames();
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/TrainedModel.java]: Both these methods are just not called anywhere? Is that why they are removed? If so 👍

**Review by davidkyle — APPROVED** [review]
LGTM. I noted some minor things and would like to see the classification labels in `LangIdentNeuralNetwork::toXContent`

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceHelpers.java [hunk 4]
```diff
@@ -20,27 +21,40 @@
...
+                                                                                                List<String> classificationLabels,
+                                                                                                @Nullable double[] classificationWeights,
+                                                                                                int numToInclude) {
 
         if (classificationLabels != null && probabilities.size() != classificationLabels.size()) {
             throw ExceptionsHelper
                 .serverError(
-                    "model returned classification probabilities of size [{}] which is not equal to classification labels size [{}]",
+                    "model returned classification probabilities of size [{}] which is not equal to classification labels size {}",
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceHelpers.java]: 🤔 possibly? I might want to switch this around a bit. It is writing out the classification labels array not the size. Woops!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceHelpers.java (hunk 2 above)
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceHelpers.java]: There is coverage through functional tests.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/TrainedModel.java (hunk 3 above)
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/TrainedModel.java]: Yes, I am cleaning up the interface. I added a bunch of stuff that I thought "all" models would need, and it turns out that that is not the case.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceHelpers.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidkyle) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceHelpers.java]: Ha

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Pull request merged**